File: dotnet/samples/Concepts/Agents/ChatCompletion_Templating.cs
// Copyright (c) Microsoft. All rights reserved.
using Microsoft.SemanticKernel;
using Microsoft.SemanticKernel.Agents;
using Microsoft.SemanticKernel.ChatCompletion;
using Microsoft.SemanticKernel.PromptTemplates.Handlebars;
using Microsoft.SemanticKernel.PromptTemplates.Liquid;

namespace Agents;

/// <summary>
/// Demonstrate parameterized template instruction for <see cref="ChatCompletionAgent"/>.
/// </summary>
public class ChatCompletion_Templating(ITestOutputHelper output) : BaseAgentsTest(output)
{
    private static readonly (string Input, string? Style)[] s_inputs =
        [
            (Input: "Home cooking is great.", Style: null),
            (Input: "Talk about world peace.", Style: "iambic pentameter"),
            (Input: "Say something about doing your best.", Style: "e. e. cummings"),
            (Input: "What do you think about having fun?", Style: "old school rap")
        ];

    [Theory]
    [InlineData(true)]
    [InlineData(false)]
    public async Task InvokeAgentWithInstructionsTemplate(bool useChatClient)
    {
        // Instruction based template always processed by KernelPromptTemplateFactory
        ChatCompletionAgent agent =
            new()
            {
                Kernel = this.CreateKernelWithChatCompletion(useChatClient, out var chatClient),
                Instructions =
                    """
                    Write a one verse poem on the requested topic in the style of {{$style}}.
                    Always state the requested style of the poem.
                    """,
                Arguments = new KernelArguments()
                {
                    {"style", "haiku"}
                }
            };

        await InvokeChatCompletionAgentWithTemplateAsync(agent);

        chatClient?.Dispose();
    }

    [Theory]
    [InlineData(true)]
    [InlineData(false)]
    public async Task InvokeAgentWithKernelTemplate(bool useChatClient)
    {
        // Default factory is KernelPromptTemplateFactory
        await InvokeChatCompletionAgentWithTemplateAsync(
            """
            Write a one verse poem on the requested topic in the style of {{$style}}.
            Always state the requested style of the poem.
            """,
            PromptTemplateConfig.SemanticKernelTemplateFormat,
            new KernelPromptTemplateFactory(),
            useChatClient);
    }

    [Theory]
    [InlineData(true)]
    [InlineData(false)]
    public async Task InvokeAgentWithHandlebarsTemplate(bool useChatClient)
    {
        await InvokeChatCompletionAgentWithTemplateAsync(
            """
            Write a one verse poem on the requested topic in the style of {{style}}.
            Always state the requested style of the poem.
            """,
            HandlebarsPromptTemplateFactory.HandlebarsTemplateFormat,
            new HandlebarsPromptTemplateFactory(),
            useChatClient);
    }

    [Theory]
    [InlineData(true)]
    [InlineData(false)]
    public async Task InvokeAgentWithLiquidTemplate(bool useChatClient)
    {
        await InvokeChatCompletionAgentWithTemplateAsync(
            """
            Write a one verse poem on the requested topic in the style of {{style}}.
            Always state the requested style of the poem.
            """,
            LiquidPromptTemplateFactory.LiquidTemplateFormat,
            new LiquidPromptTemplateFactory(),
            useChatClient);
    }

    private async Task InvokeChatCompletionAgentWithTemplateAsync(
        string instructionTemplate,
        string templateFormat,
        IPromptTemplateFactory templateFactory,
        bool useChatClient)
    {
        // Define the agent
        PromptTemplateConfig templateConfig =
            new()
            {
                Template = instructionTemplate,
                TemplateFormat = templateFormat,
            };
        ChatCompletionAgent agent =
            new(templateConfig, templateFactory)
            {
                Kernel = this.CreateKernelWithChatCompletion(useChatClient, out var chatClient),
                Arguments = new KernelArguments()
                {
                    {"style", "haiku"}
                }
            };

        await InvokeChatCompletionAgentWithTemplateAsync(agent);

        chatClient?.Dispose();
    }

    private async Task InvokeChatCompletionAgentWithTemplateAsync(ChatCompletionAgent agent)
    {
        ChatHistory chat = [];

        foreach ((string input, string? style) in s_inputs)
        {
            // Add input to chat
            ChatMessageContent request = new(AuthorRole.User, input);
            this.WriteAgentChatMessage(request);

            KernelArguments? arguments = null;

            if (!string.IsNullOrWhiteSpace(style))
            {
                // Override style template parameter
                arguments = new() { { "style", style } };
            }

            // Process agent response
            await foreach (ChatMessageContent message in agent.InvokeAsync(request, options: new() { KernelArguments = arguments }))
            {
                chat.Add(message);
                this.WriteAgentChatMessage(message);
            }
        }
    }
}

File: python/samples/concepts/agents/chat_completion_templating.py
# Copyright (c) Microsoft. All rights reserved.

import asyncio

from semantic_kernel.agents import ChatCompletionAgent
from semantic_kernel.connectors.ai.open_ai import AzureChatCompletion
from semantic_kernel.contents.chat_history import ChatHistory
from semantic_kernel.contents.utils.author_role import AuthorRole
from semantic_kernel.functions.kernel_arguments import KernelArguments
from semantic_kernel.kernel import Kernel
from semantic_kernel.prompt_template.prompt_template_config import PromptTemplateConfig

###################################################################
# The following sample demonstrates how to configure the auto     #
# function invocation filter with use of a ChatCompletionAgent.   #
###################################################################


def _create_kernel_with_chat_completionand_filter(service_id: str) -> Kernel:
    """A helper function to create a kernel with a chat completion service and a filter."""
    kernel = Kernel()
    kernel.add_service(AzureChatCompletion(service_id=service_id))
    return kernel


# A helper method to invoke the agent with the user input
async def invoke_agent(agent: ChatCompletionAgent, input: str, chat_history: ChatHistory) -> None:
    """Invoke the agent with the user input."""
    chat_history.add_user_message(input)
    print(f"# {AuthorRole.USER}: '{input}'")

    async for message in agent.invoke(chat_history):
        chat_history.add_message(message)
        print(message.content)


async def main():
    service_id = "agent"

    # Create the kernel used by the chat completion agent
    kernel = _create_kernel_with_chat_completionand_filter(service_id=service_id)

    # Create the agent
    # agent = ChatCompletionAgent(
    #     service_id=service_id,
    #     kernel=kernel,
    #     instructions="""
    #     Write a one verse poem on the requested topic in the style of {{$style}}.
    #     Always state the requested style of the poem.
    #     """,
    #     arguments=KernelArguments(style="Haiku"),
    # )

    # h_prompt_template_config = PromptTemplateConfig(
    #     template="""
    #     Write a one verse poem on the requested topic in the style of {{style}}.
    #     Always state the requested style of the poem.
    #     """,
    #     template_format="handlebars",
    # )

    # handlebars_agent = ChatCompletionAgent(
    #     service_id=service_id,
    #     kernel=kernel,
    #     arguments=KernelArguments(style="Haiku"),
    #     prompt_template_config=h_prompt_template_config,
    # )

    jinja2_agent = ChatCompletionAgent(
        service_id=service_id,
        kernel=kernel,
        arguments=KernelArguments(style="Haiku"),
        prompt_template_config=PromptTemplateConfig(
            template="""
            Write a one verse poem on the requested topic in the style of {{style}}.
            Always state the requested style of the poem.
            """,
            template_format="jinja2",
        ),
    )

    # Define the chat history
    chat = ChatHistory()

    # Respond to user input
    await invoke_agent(agent=jinja2_agent, input="Home cooking is great.", chat_history=chat)


if __name__ == "__main__":
    asyncio.run(main())
